op-service/sources/l1_beacon_client.go

```diff
@@ -0,0 +1,157 @@
+// checked for validity.
+func (cl *L1BeaconClient) GetBlobSidecars(ctx context.Context, ref eth.L1BlockRef, hashes []eth.IndexedBlobHash) ([]*eth.BlobSidecar, error) {
+	if len(hashes) == 0 {
+		return []*eth.BlobSidecar{}, nil
+	}
+	slotFn, err := cl.GetTimeToSlotFn(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to get time to slot function", err)
+	}
+	slot := slotFn(ref.Time)
+
+	builder := strings.Builder{}
+	builder.WriteString("eth/v1/beacon/blob_sidecars/")
+	builder.WriteString(strconv.FormatUint(slot, 10))
+	for i := range hashes {
+		if i == 0 {
+			builder.WriteString("?indices=")
+		} else {
+			builder.WriteString("&indices=")
+		}
+		builder.WriteString(strconv.FormatUint(hashes[i].Index, 10))
+	}
+
+	var resp eth.APIGetBlobSidecarsResponse
+	if err := cl.apiReq(ctx, &resp, builder.String()); err != nil {
+		return nil, fmt.Errorf("%w: failed to fetch blob sidecars for slot %v block %v", err, slot, ref)
+	}
+	if len(hashes) != len(resp.Data) {
+		return nil, fmt.Errorf("expected %v sidecars but got %v", len(hashes), len(resp.Data))
+	}
+
+	return resp.Data, nil
+}
+
+// GetBlobs fetches blobs that were confirmed in the specified L1 block with the given indexed
+// hashes. The order of the returned blobs will match the order of `hashes`.  Confirms each
+// blob's validity by checking its proof against the commitment, and confirming the commitment
+// hashes to the expected value. Returns error if any blob is found invalid.
+func (cl *L1BeaconClient) GetBlobs(ctx context.Context, ref eth.L1BlockRef, hashes []eth.IndexedBlobHash) ([]*eth.Blob, error) {
+	blobSidecars, err := cl.GetBlobSidecars(ctx, ref, hashes)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to get blob sidecars for L1BlockRef %s", err, ref)
+	}
+	return blobsFromSidecars(blobSidecars, hashes)
+}
+
+func blobsFromSidecars(blobSidecars []*eth.BlobSidecar, hashes []eth.IndexedBlobHash) ([]*eth.Blob, error) {
+	out := make([]*eth.Blob, len(hashes))
+	for i, ih := range hashes {
+		// The beacon node api makes no guarantees on order of the returned blob sidecars, so
+		// search for the sidecar that matches the current indexed hash to ensure blobs are
+		// returned in the same order.
+		var sidecar *eth.BlobSidecar
+		for _, sc := range blobSidecars {
+			if uint64(sc.Index) == ih.Index {
+				sidecar = sc
+				break
+			}
+		}
+		if sidecar == nil {
+			return nil, fmt.Errorf("no blob in response matches desired index: %v", ih.Index)
+		}
+
+		// make sure the blob's kzg commitment hashes to the expected value
+		hash := eth.KZGToVersionedHash(kzg4844.Commitment(sidecar.KZGCommitment))
+		if hash != ih.Hash {
+			return nil, fmt.Errorf("expected hash %s for blob at index %d but got %s", ih.Hash, ih.Index, hash)
+		}
+
+		// confirm blob data is valid by verifying its proof against the commitment
+		if err := eth.VerifyBlobProof(&sidecar.Blob, kzg4844.Commitment(sidecar.KZGCommitment), kzg4844.Proof(sidecar.KZGProof)); err != nil {
```

Comment:
Is there a case in which we want to skip the verify? I know that other places we can omit verification with a config option

Reply:
I feel the throughput of blobs is low enough that it's best to just always do the check rather than over-complicate but open to other opinions.